{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc7b6337",
   "metadata": {},
   "source": [
    "## This notebook created by Jason Miller https://github.com/ShepherdCode/ShepherdML/blob/master/Localization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c80a152",
   "metadata": {
    "id": "6c80a152"
   },
   "source": [
    "## GenCode\n",
    "https://www.gencodegenes.org/human/release_42.html\n",
    "\n",
    "This notebook runs before Train/Test Split.\n",
    "\n",
    "This notebook reduces our GenCode sequence files to\n",
    "\n",
    "    genes with RCI values in LncAtlas\n",
    "    transcripts whose GenCode annotation looks good\n",
    "\n",
    "\n",
    "\n",
    "Human sequence in FASTA:\n",
    "\n",
    "    genomic DNA\n",
    "    gene cDNA (with introns)\n",
    "    transcript CDS (without introns)\n",
    "    noncoding ncRNA\n",
    "    Human gene annotation in GFF\n",
    "    *****\n",
    "    gencode.v42.annotation.gff3\n",
    "    gencode.v42.lncRNA_transcripts.fa\n",
    "    gencode.v42.pc_transcripts.fa\n",
    "    ****"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d66efe1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "jiYEc_j_zJfQ",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jiYEc_j_zJfQ",
    "outputId": "faeb1d8e-4974-4ca9-a4a3-0ca6fafcd04e"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:10:20.916684\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "43c31b19",
   "metadata": {
    "id": "43c31b19"
   },
   "outputs": [],
   "source": [
    "DATA_DIR = \"../../data/GenCode/\"\n",
    "ATLAS_DIR = \"../../data/LncAtlas/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7ae29f96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gencode.v42.annotation.gff3',\n",
       " 'gencode.v42.lncRNA_transcripts.fa',\n",
       " 'gencode.v42.pc_transcripts.fa']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(DATA_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "NoFbNkY4yLF7",
   "metadata": {
    "id": "NoFbNkY4yLF7"
   },
   "outputs": [],
   "source": [
    "# GenCode inputs\n",
    "\n",
    "ANNOTATION = 'gencode.v42.annotation.gff3'\n",
    "NONCODING_SEQUENCE = 'gencode.v42.lncRNA_transcripts.fa'\n",
    "CODING_SEQUENCE = 'gencode.v42.pc_transcripts.fa'\n",
    "# GenCode outputs\n",
    "CODING_CSV = 'gencode.v42.pc_transcripts.csv'\n",
    "NONCODING_CSV = 'gencode.v42.lncRNA_transcripts.csv'\n",
    "# Atlas inputs\n",
    "ATLAS_FILE = 'lncATLAS_all_data_RCI.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vLgSe2LeBX7Z",
   "metadata": {
    "id": "vLgSe2LeBX7Z"
   },
   "outputs": [],
   "source": [
    "# dff = pd.read_csv(DATA_DIR+ANNOTATION)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "79290b90",
   "metadata": {
    "id": "79290b90"
   },
   "outputs": [],
   "source": [
    "class fasta_reader():\n",
    "    '''\n",
    "    Parser for human transcripts FASTA file from GenCode.\n",
    "    '''\n",
    "    def __init__(self,infile,outfile,biotype):\n",
    "        '''\n",
    "        Biotype should reflect the filename: either 'pc' or 'lncRNA'.\n",
    "        '''\n",
    "        self.infile = infile\n",
    "        self.outfile = outfile\n",
    "        self.biotype = biotype\n",
    "        self.FASTA_DEFCHAR = '>'  # signals a defline = definition line\n",
    "        self.count_in = 0\n",
    "        self.count_out = 0\n",
    "        self.allow_genes = None\n",
    "        self.allow_transcripts = None\n",
    "        self.headers='transcript_id,gene_id,biotype,length,sequence\\n'\n",
    "    def allow_these_genes(self,genes:set):\n",
    "        self.allow_genes = genes\n",
    "    def allow_these_transcripts(self,trans:set):\n",
    "        self.allow_transcripts = trans\n",
    "    def print_one_sequence(self,handle,tran,gene,seq):\n",
    "        allow_genes = self.allow_genes\n",
    "        allow_trans = self.allow_transcripts\n",
    "        if seq is not None:\n",
    "            # sequence is None when we encounter the first defline\n",
    "            if allow_genes is None or gene in allow_genes:\n",
    "                if allow_trans is None or tran in allow_trans:\n",
    "                    biotype = self.biotype\n",
    "                    length = str(len(seq))\n",
    "                    outstr = ','.join((tran,gene,biotype,length,seq))\n",
    "                    handle.write(outstr+'\\n')\n",
    "                    self.count_out += 1\n",
    "    def fasta_to_csv(self):\n",
    "        with open(self.outfile,'w') as handle:\n",
    "            handle.write(self.headers)\n",
    "            with open(self.infile,'r') as fasta:\n",
    "                transcript_id = None\n",
    "                gene_id = None\n",
    "                next_seq = None\n",
    "                for line in fasta:\n",
    "                    if line[0]==self.FASTA_DEFCHAR:\n",
    "                        self.count_in += 1\n",
    "                        # The defline starts with '>'\n",
    "                        # The defline has fields separated by vertical bar\n",
    "                        # Wrap up the previous sequence before moving on to the next.\n",
    "                        self.print_one_sequence(handle,transcript_id,gene_id,next_seq)\n",
    "                        tokens = line.split('|')\n",
    "                        transcript_id = tokens[0][1:] # chop off '>'\n",
    "                        gene_id = tokens[1]\n",
    "                        version_index=gene_id.find('.')\n",
    "                        if version_index>=0:\n",
    "                            # chop off version number, as in ENSG00000198888.2\n",
    "                            gene_id = gene_id[:version_index]\n",
    "                        next_seq = \"\"   # get ready for one to many sequence lines\n",
    "                    else:\n",
    "                        # In FASTA format, one sequence may continue to next line\n",
    "                        next_seq = next_seq + line.strip()\n",
    "            self.print_one_sequence(handle,transcript_id,gene_id,next_seq)\n",
    "        print(\" Input sequences: %d\"%self.count_in)\n",
    "        print(\"Output sequences: %d\"%self.count_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7nrdoMhezDRQ",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "7nrdoMhezDRQ",
    "outputId": "6602b36e-37db-4e61-8e45-9b8aaee42295"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:14:46.446961\n",
      "Atlas good genes: 25172\n"
     ]
    }
   ],
   "source": [
    "def load_atlas_genes(filepath):\n",
    "    genes = set()\n",
    "    with open (filepath,'r') as handle:\n",
    "        header = None\n",
    "        for row in handle:\n",
    "            if header is None:\n",
    "                header = row\n",
    "            else:\n",
    "                fields = row.split(',')\n",
    "                gene_id = fields[0]\n",
    "                value = fields[3]\n",
    "                if (value != 'NA'):\n",
    "                    genes.add(gene_id)  # set removes dupes\n",
    "    return genes\n",
    "print(datetime.now())\n",
    "atlas_genes = load_atlas_genes(ATLAS_DIR+ATLAS_FILE)\n",
    "print('Atlas good genes:', len(atlas_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "zrXwy8sFjiHM",
   "metadata": {
    "collapsed": true,
    "id": "zrXwy8sFjiHM"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000003056',\n",
       " 'ENSG00000156564',\n",
       " 'ENSG00000108448',\n",
       " 'ENSG00000178199',\n",
       " 'ENSG00000228560',\n",
       " 'ENSG00000271533',\n",
       " 'ENSG00000188626',\n",
       " 'ENSG00000123610',\n",
       " 'ENSG00000136274',\n",
       " 'ENSG00000258957',\n",
       " 'ENSG00000136758',\n",
       " 'ENSG00000107779',\n",
       " 'ENSG00000162086',\n",
       " 'ENSG00000113448',\n",
       " 'ENSG00000152779',\n",
       " 'ENSG00000035687',\n",
       " 'ENSG00000137404',\n",
       " 'ENSG00000282572',\n",
       " 'ENSG00000123600',\n",
       " 'ENSG00000258102',\n",
       " 'ENSG00000139767',\n",
       " 'ENSG00000235192',\n",
       " 'ENSG00000253649',\n",
       " 'ENSG00000231749',\n",
       " 'ENSG00000176092',\n",
       " 'ENSG00000188677',\n",
       " 'ENSG00000120915',\n",
       " 'ENSG00000281348',\n",
       " 'ENSG00000256577',\n",
       " 'ENSG00000280422',\n",
       " 'ENSG00000138074',\n",
       " 'ENSG00000279141',\n",
       " 'ENSG00000162704',\n",
       " 'ENSG00000249645',\n",
       " 'ENSG00000185917',\n",
       " 'ENSG00000204673',\n",
       " 'ENSG00000103512',\n",
       " 'ENSG00000279781',\n",
       " 'ENSG00000271840',\n",
       " 'ENSG00000027075',\n",
       " 'ENSG00000272328',\n",
       " 'ENSG00000280789',\n",
       " 'ENSG00000089053',\n",
       " 'ENSG00000267470',\n",
       " 'ENSG00000144452',\n",
       " 'ENSG00000047346',\n",
       " 'ENSG00000283075',\n",
       " 'ENSG00000196189',\n",
       " 'ENSG00000178685',\n",
       " 'ENSG00000272343',\n",
       " 'ENSG00000166548',\n",
       " 'ENSG00000277476',\n",
       " 'ENSG00000095787',\n",
       " 'ENSG00000188976',\n",
       " 'ENSG00000250682',\n",
       " 'ENSG00000115738',\n",
       " 'ENSG00000183844',\n",
       " 'ENSG00000139433',\n",
       " 'ENSG00000228951',\n",
       " 'ENSG00000122490',\n",
       " 'ENSG00000279333',\n",
       " 'ENSG00000106028',\n",
       " 'ENSG00000196361',\n",
       " 'ENSG00000227706',\n",
       " 'ENSG00000227740',\n",
       " 'ENSG00000267462',\n",
       " 'ENSG00000178460',\n",
       " 'ENSG00000173653',\n",
       " 'ENSG00000179859',\n",
       " 'ENSG00000279233',\n",
       " 'ENSG00000131504',\n",
       " 'ENSG00000280424',\n",
       " 'ENSG00000188655',\n",
       " 'ENSG00000166582',\n",
       " 'ENSG00000162408',\n",
       " 'ENSG00000279149',\n",
       " 'ENSG00000131149',\n",
       " 'ENSG00000219445',\n",
       " 'ENSG00000109534',\n",
       " 'ENSG00000130382',\n",
       " 'ENSG00000261610',\n",
       " 'ENSG00000258038',\n",
       " 'ENSG00000182578',\n",
       " 'ENSG00000123395',\n",
       " 'ENSG00000236255',\n",
       " 'ENSG00000175482',\n",
       " 'ENSG00000160886',\n",
       " 'ENSG00000255046',\n",
       " 'ENSG00000067057',\n",
       " 'ENSG00000125430',\n",
       " 'ENSG00000198740',\n",
       " 'ENSG00000196313',\n",
       " 'ENSG00000037241',\n",
       " 'ENSG00000228278',\n",
       " 'ENSG00000229896',\n",
       " 'ENSG00000120437',\n",
       " 'ENSG00000185880',\n",
       " 'ENSG00000131389',\n",
       " 'ENSG00000142453',\n",
       " 'ENSG00000226604',\n",
       " 'ENSG00000261542',\n",
       " 'ENSG00000157680',\n",
       " 'ENSG00000124126',\n",
       " 'ENSG00000182359',\n",
       " 'ENSG00000204472',\n",
       " 'ENSG00000174951',\n",
       " 'ENSG00000275457',\n",
       " 'ENSG00000174231',\n",
       " 'ENSG00000149346',\n",
       " 'ENSG00000259970',\n",
       " 'ENSG00000161904',\n",
       " 'ENSG00000101076',\n",
       " 'ENSG00000011304',\n",
       " 'ENSG00000261118',\n",
       " 'ENSG00000177045',\n",
       " 'ENSG00000230565',\n",
       " 'ENSG00000183495',\n",
       " 'ENSG00000141447',\n",
       " 'ENSG00000155313',\n",
       " 'ENSG00000267493',\n",
       " 'ENSG00000198821',\n",
       " 'ENSG00000088992',\n",
       " 'ENSG00000224905',\n",
       " 'ENSG00000234409',\n",
       " 'ENSG00000163520',\n",
       " 'ENSG00000198039',\n",
       " 'ENSG00000205927',\n",
       " 'ENSG00000134551',\n",
       " 'ENSG00000231439',\n",
       " 'ENSG00000157765',\n",
       " 'ENSG00000204642',\n",
       " 'ENSG00000007516',\n",
       " 'ENSG00000266236',\n",
       " 'ENSG00000225815',\n",
       " 'ENSG00000274400',\n",
       " 'ENSG00000119715',\n",
       " 'ENSG00000226751',\n",
       " 'ENSG00000228952',\n",
       " 'ENSG00000136280',\n",
       " 'ENSG00000101825',\n",
       " 'ENSG00000173614',\n",
       " 'ENSG00000280798',\n",
       " 'ENSG00000132874',\n",
       " 'ENSG00000266934',\n",
       " 'ENSG00000266256',\n",
       " 'ENSG00000005801',\n",
       " 'ENSG00000178935',\n",
       " 'ENSG00000183273',\n",
       " 'ENSG00000114354',\n",
       " 'ENSG00000226383',\n",
       " 'ENSG00000273551',\n",
       " 'ENSG00000204438',\n",
       " 'ENSG00000186474',\n",
       " 'ENSG00000080561',\n",
       " 'ENSG00000173011',\n",
       " 'ENSG00000160882',\n",
       " 'ENSG00000183389',\n",
       " 'ENSG00000275963',\n",
       " 'ENSG00000215447',\n",
       " 'ENSG00000255136',\n",
       " 'ENSG00000149091',\n",
       " 'ENSG00000130812',\n",
       " 'ENSG00000099866',\n",
       " 'ENSG00000181143',\n",
       " 'ENSG00000101333',\n",
       " 'ENSG00000135622',\n",
       " 'ENSG00000197249',\n",
       " 'ENSG00000229862',\n",
       " 'ENSG00000251143',\n",
       " 'ENSG00000164808',\n",
       " 'ENSG00000166762',\n",
       " 'ENSG00000198879',\n",
       " 'ENSG00000261087',\n",
       " 'ENSG00000124839',\n",
       " 'ENSG00000207955',\n",
       " 'ENSG00000225127',\n",
       " 'ENSG00000071203',\n",
       " 'ENSG00000273032',\n",
       " 'ENSG00000143258',\n",
       " 'ENSG00000166685',\n",
       " 'ENSG00000218018',\n",
       " 'ENSG00000253642',\n",
       " 'ENSG00000112195',\n",
       " 'ENSG00000091127',\n",
       " 'ENSG00000260281',\n",
       " 'ENSG00000273155',\n",
       " 'ENSG00000269514',\n",
       " 'ENSG00000136267',\n",
       " 'ENSG00000139926',\n",
       " 'ENSG00000108352',\n",
       " 'ENSG00000100354',\n",
       " 'ENSG00000197885',\n",
       " 'ENSG00000213024',\n",
       " 'ENSG00000125458',\n",
       " 'ENSG00000104903',\n",
       " 'ENSG00000185860',\n",
       " 'ENSG00000165794',\n",
       " 'ENSG00000228437',\n",
       " 'ENSG00000229081',\n",
       " 'ENSG00000170927',\n",
       " 'ENSG00000110619',\n",
       " 'ENSG00000177485',\n",
       " 'ENSG00000258588',\n",
       " 'ENSG00000112146',\n",
       " 'ENSG00000158517',\n",
       " 'ENSG00000265590',\n",
       " 'ENSG00000271383',\n",
       " 'ENSG00000169242',\n",
       " 'ENSG00000153902',\n",
       " 'ENSG00000175445',\n",
       " 'ENSG00000227502',\n",
       " 'ENSG00000102802',\n",
       " 'ENSG00000180269',\n",
       " 'ENSG00000232311',\n",
       " 'ENSG00000184515',\n",
       " 'ENSG00000253251',\n",
       " 'ENSG00000184682',\n",
       " 'ENSG00000099365',\n",
       " 'ENSG00000237594',\n",
       " 'ENSG00000128045',\n",
       " 'ENSG00000203952',\n",
       " 'ENSG00000248167',\n",
       " 'ENSG00000279453',\n",
       " 'ENSG00000124299',\n",
       " 'ENSG00000171169',\n",
       " 'ENSG00000140400',\n",
       " 'ENSG00000257207',\n",
       " 'ENSG00000139572',\n",
       " 'ENSG00000088812',\n",
       " 'ENSG00000164077',\n",
       " 'ENSG00000249464',\n",
       " 'ENSG00000279488',\n",
       " 'ENSG00000170881',\n",
       " 'ENSG00000139304',\n",
       " 'ENSG00000228022',\n",
       " 'ENSG00000237857',\n",
       " 'ENSG00000250170',\n",
       " 'ENSG00000235151',\n",
       " 'ENSG00000248161',\n",
       " 'ENSG00000124194',\n",
       " 'ENSG00000260751',\n",
       " 'ENSG00000044524',\n",
       " 'ENSG00000171604',\n",
       " 'ENSG00000005448',\n",
       " 'ENSG00000188039',\n",
       " 'ENSG00000187475',\n",
       " 'ENSG00000109670',\n",
       " 'ENSG00000070501',\n",
       " 'ENSG00000245888',\n",
       " 'ENSG00000159882',\n",
       " 'ENSG00000143512',\n",
       " 'ENSG00000167618',\n",
       " 'ENSG00000099721',\n",
       " 'ENSG00000157224',\n",
       " 'ENSG00000100335',\n",
       " 'ENSG00000115590',\n",
       " 'ENSG00000204632',\n",
       " 'ENSG00000075388',\n",
       " 'ENSG00000110244',\n",
       " 'ENSG00000039139',\n",
       " 'ENSG00000166432',\n",
       " 'ENSG00000035141',\n",
       " 'ENSG00000117477',\n",
       " 'ENSG00000224322',\n",
       " 'ENSG00000154917',\n",
       " 'ENSG00000205464',\n",
       " 'ENSG00000245849',\n",
       " 'ENSG00000117877',\n",
       " 'ENSG00000221823',\n",
       " 'ENSG00000066557',\n",
       " 'ENSG00000182372',\n",
       " 'ENSG00000233654',\n",
       " 'ENSG00000124588',\n",
       " 'ENSG00000143183',\n",
       " 'ENSG00000273148',\n",
       " 'ENSG00000100100',\n",
       " 'ENSG00000250988',\n",
       " 'ENSG00000170242',\n",
       " 'ENSG00000277858',\n",
       " 'ENSG00000142700',\n",
       " 'ENSG00000144034',\n",
       " 'ENSG00000183813',\n",
       " 'ENSG00000268472',\n",
       " 'ENSG00000163872',\n",
       " 'ENSG00000171428',\n",
       " 'ENSG00000281912',\n",
       " 'ENSG00000100281',\n",
       " 'ENSG00000168803',\n",
       " 'ENSG00000255408',\n",
       " 'ENSG00000268520',\n",
       " 'ENSG00000261845',\n",
       " 'ENSG00000048028',\n",
       " 'ENSG00000279355',\n",
       " 'ENSG00000280137',\n",
       " 'ENSG00000108312',\n",
       " 'ENSG00000130227',\n",
       " 'ENSG00000100483',\n",
       " 'ENSG00000099341',\n",
       " 'ENSG00000153246',\n",
       " 'ENSG00000249201',\n",
       " 'ENSG00000125347',\n",
       " 'ENSG00000117054',\n",
       " 'ENSG00000225393',\n",
       " 'ENSG00000104332',\n",
       " 'ENSG00000280355',\n",
       " 'ENSG00000234722',\n",
       " 'ENSG00000270673',\n",
       " 'ENSG00000279474',\n",
       " 'ENSG00000115947',\n",
       " 'ENSG00000225234',\n",
       " 'ENSG00000112592',\n",
       " 'ENSG00000134294',\n",
       " 'ENSG00000168754',\n",
       " 'ENSG00000099139',\n",
       " 'ENSG00000137204',\n",
       " 'ENSG00000105388',\n",
       " 'ENSG00000204361',\n",
       " 'ENSG00000165475',\n",
       " 'ENSG00000272983',\n",
       " 'ENSG00000103005',\n",
       " 'ENSG00000106541',\n",
       " 'ENSG00000067221',\n",
       " 'ENSG00000107281',\n",
       " 'ENSG00000142733',\n",
       " 'ENSG00000184500',\n",
       " 'ENSG00000185477',\n",
       " 'ENSG00000185504',\n",
       " 'ENSG00000116350',\n",
       " 'ENSG00000242259',\n",
       " 'ENSG00000273906',\n",
       " 'ENSG00000144848',\n",
       " 'ENSG00000197780',\n",
       " 'ENSG00000228115',\n",
       " 'ENSG00000183049',\n",
       " 'ENSG00000158042',\n",
       " 'ENSG00000160678',\n",
       " 'ENSG00000258311',\n",
       " 'ENSG00000180917',\n",
       " 'ENSG00000167654',\n",
       " 'ENSG00000258976',\n",
       " 'ENSG00000175895',\n",
       " 'ENSG00000118246',\n",
       " 'ENSG00000277010',\n",
       " 'ENSG00000180488',\n",
       " 'ENSG00000157827',\n",
       " 'ENSG00000072818',\n",
       " 'ENSG00000198862',\n",
       " 'ENSG00000266978',\n",
       " 'ENSG00000198363',\n",
       " 'ENSG00000277352',\n",
       " 'ENSG00000279860',\n",
       " 'ENSG00000166889',\n",
       " 'ENSG00000234076',\n",
       " 'ENSG00000156042',\n",
       " 'ENSG00000179454',\n",
       " 'ENSG00000110756',\n",
       " 'ENSG00000273523',\n",
       " 'ENSG00000240211',\n",
       " 'ENSG00000214353',\n",
       " 'ENSG00000077080',\n",
       " 'ENSG00000160948',\n",
       " 'ENSG00000106443',\n",
       " 'ENSG00000163217',\n",
       " 'ENSG00000214063',\n",
       " 'ENSG00000139915',\n",
       " 'ENSG00000152942',\n",
       " 'ENSG00000144406',\n",
       " 'ENSG00000072571',\n",
       " 'ENSG00000119915',\n",
       " 'ENSG00000136521',\n",
       " 'ENSG00000111885',\n",
       " 'ENSG00000100503',\n",
       " 'ENSG00000100485',\n",
       " 'ENSG00000157999',\n",
       " 'ENSG00000147789',\n",
       " 'ENSG00000172059',\n",
       " 'ENSG00000276863',\n",
       " 'ENSG00000164062',\n",
       " 'ENSG00000203635',\n",
       " 'ENSG00000251216',\n",
       " 'ENSG00000095574',\n",
       " 'ENSG00000260583',\n",
       " 'ENSG00000141956',\n",
       " 'ENSG00000235688',\n",
       " 'ENSG00000275496',\n",
       " 'ENSG00000141736',\n",
       " 'ENSG00000188674',\n",
       " 'ENSG00000187013',\n",
       " 'ENSG00000183889',\n",
       " 'ENSG00000249082',\n",
       " 'ENSG00000110680',\n",
       " 'ENSG00000260093',\n",
       " 'ENSG00000140691',\n",
       " 'ENSG00000138083',\n",
       " 'ENSG00000272182',\n",
       " 'ENSG00000135502',\n",
       " 'ENSG00000279829',\n",
       " 'ENSG00000216921',\n",
       " 'ENSG00000173273',\n",
       " 'ENSG00000261496',\n",
       " 'ENSG00000107954',\n",
       " 'ENSG00000058729',\n",
       " 'ENSG00000187609',\n",
       " 'ENSG00000135750',\n",
       " 'ENSG00000242498',\n",
       " 'ENSG00000203697',\n",
       " 'ENSG00000144233',\n",
       " 'ENSG00000239906',\n",
       " 'ENSG00000224975',\n",
       " 'ENSG00000242797',\n",
       " 'ENSG00000255104',\n",
       " 'ENSG00000160305',\n",
       " 'ENSG00000279145',\n",
       " 'ENSG00000257337',\n",
       " 'ENSG00000274267',\n",
       " 'ENSG00000065809',\n",
       " 'ENSG00000278727',\n",
       " 'ENSG00000152214',\n",
       " 'ENSG00000196620',\n",
       " 'ENSG00000102871',\n",
       " 'ENSG00000182179',\n",
       " 'ENSG00000226824',\n",
       " 'ENSG00000204092',\n",
       " 'ENSG00000186834',\n",
       " 'ENSG00000113302',\n",
       " 'ENSG00000256802',\n",
       " 'ENSG00000237187',\n",
       " 'ENSG00000176153',\n",
       " 'ENSG00000253477',\n",
       " 'ENSG00000260757',\n",
       " 'ENSG00000115423',\n",
       " 'ENSG00000108733',\n",
       " 'ENSG00000274993',\n",
       " 'ENSG00000251323',\n",
       " 'ENSG00000129682',\n",
       " 'ENSG00000158435',\n",
       " 'ENSG00000162398',\n",
       " 'ENSG00000160396',\n",
       " 'ENSG00000260448',\n",
       " 'ENSG00000179111',\n",
       " 'ENSG00000236525',\n",
       " 'ENSG00000050405',\n",
       " 'ENSG00000187690',\n",
       " 'ENSG00000280073',\n",
       " 'ENSG00000178860',\n",
       " 'ENSG00000130669',\n",
       " 'ENSG00000188816',\n",
       " 'ENSG00000198624',\n",
       " 'ENSG00000180881',\n",
       " 'ENSG00000171606',\n",
       " 'ENSG00000253873',\n",
       " 'ENSG00000165983',\n",
       " 'ENSG00000151422',\n",
       " 'ENSG00000101972',\n",
       " 'ENSG00000072954',\n",
       " 'ENSG00000267194',\n",
       " 'ENSG00000172469',\n",
       " 'ENSG00000255750',\n",
       " 'ENSG00000172375',\n",
       " 'ENSG00000264301',\n",
       " 'ENSG00000063015',\n",
       " 'ENSG00000164045',\n",
       " 'ENSG00000272097',\n",
       " 'ENSG00000125484',\n",
       " 'ENSG00000279882',\n",
       " 'ENSG00000223776',\n",
       " 'ENSG00000254997',\n",
       " 'ENSG00000099804',\n",
       " 'ENSG00000129317',\n",
       " 'ENSG00000136883',\n",
       " 'ENSG00000255317',\n",
       " 'ENSG00000251136',\n",
       " 'ENSG00000263264',\n",
       " 'ENSG00000196305',\n",
       " 'ENSG00000179144',\n",
       " 'ENSG00000275638',\n",
       " 'ENSG00000279725',\n",
       " 'ENSG00000267939',\n",
       " 'ENSG00000004848',\n",
       " 'ENSG00000141503',\n",
       " 'ENSG00000213240',\n",
       " 'ENSG00000002834',\n",
       " 'ENSG00000096384',\n",
       " 'ENSG00000254237',\n",
       " 'ENSG00000234378',\n",
       " 'ENSG00000172733',\n",
       " 'ENSG00000111732',\n",
       " 'ENSG00000214941',\n",
       " 'ENSG00000260094',\n",
       " 'ENSG00000141480',\n",
       " 'ENSG00000171560',\n",
       " 'ENSG00000180233',\n",
       " 'ENSG00000272192',\n",
       " 'ENSG00000283000',\n",
       " 'ENSG00000100243',\n",
       " 'ENSG00000271538',\n",
       " 'ENSG00000152061',\n",
       " 'ENSG00000010256',\n",
       " 'ENSG00000145428',\n",
       " 'ENSG00000153827',\n",
       " 'ENSG00000230454',\n",
       " 'ENSG00000108370',\n",
       " 'ENSG00000023191',\n",
       " 'ENSG00000173269',\n",
       " 'ENSG00000114757',\n",
       " 'ENSG00000099817',\n",
       " 'ENSG00000204160',\n",
       " 'ENSG00000216490',\n",
       " 'ENSG00000099800',\n",
       " 'ENSG00000276232',\n",
       " 'ENSG00000118922',\n",
       " 'ENSG00000185900',\n",
       " 'ENSG00000110002',\n",
       " 'ENSG00000108474',\n",
       " 'ENSG00000259345',\n",
       " 'ENSG00000130433',\n",
       " 'ENSG00000141096',\n",
       " 'ENSG00000105143',\n",
       " 'ENSG00000131435',\n",
       " 'ENSG00000105808',\n",
       " 'ENSG00000144036',\n",
       " 'ENSG00000167769',\n",
       " 'ENSG00000153982',\n",
       " 'ENSG00000165434',\n",
       " 'ENSG00000198920',\n",
       " 'ENSG00000224536',\n",
       " 'ENSG00000273259',\n",
       " 'ENSG00000173214',\n",
       " 'ENSG00000101361',\n",
       " 'ENSG00000130222',\n",
       " 'ENSG00000141485',\n",
       " 'ENSG00000011523',\n",
       " 'ENSG00000174842',\n",
       " 'ENSG00000279162',\n",
       " 'ENSG00000163961',\n",
       " 'ENSG00000234389',\n",
       " 'ENSG00000204386',\n",
       " 'ENSG00000167600',\n",
       " 'ENSG00000233559',\n",
       " 'ENSG00000124334',\n",
       " 'ENSG00000135316',\n",
       " 'ENSG00000108622',\n",
       " 'ENSG00000169223',\n",
       " 'ENSG00000196821',\n",
       " 'ENSG00000109111',\n",
       " 'ENSG00000244479',\n",
       " 'ENSG00000076662',\n",
       " 'ENSG00000185519',\n",
       " 'ENSG00000278448',\n",
       " 'ENSG00000279667',\n",
       " 'ENSG00000118096',\n",
       " 'ENSG00000064787',\n",
       " 'ENSG00000101144',\n",
       " 'ENSG00000172137',\n",
       " 'ENSG00000056998',\n",
       " 'ENSG00000117480',\n",
       " 'ENSG00000274515',\n",
       " 'ENSG00000222011',\n",
       " 'ENSG00000167555',\n",
       " 'ENSG00000154832',\n",
       " 'ENSG00000235578',\n",
       " 'ENSG00000122376',\n",
       " 'ENSG00000183060',\n",
       " 'ENSG00000186792',\n",
       " 'ENSG00000244625',\n",
       " 'ENSG00000197405',\n",
       " 'ENSG00000160957',\n",
       " 'ENSG00000260423',\n",
       " 'ENSG00000242590',\n",
       " 'ENSG00000181061',\n",
       " 'ENSG00000280228',\n",
       " 'ENSG00000243179',\n",
       " 'ENSG00000239900',\n",
       " 'ENSG00000269720',\n",
       " 'ENSG00000275023',\n",
       " 'ENSG00000070018',\n",
       " 'ENSG00000148291',\n",
       " 'ENSG00000124145',\n",
       " 'ENSG00000235750',\n",
       " 'ENSG00000119004',\n",
       " 'ENSG00000280372',\n",
       " 'ENSG00000166710',\n",
       " 'ENSG00000116031',\n",
       " 'ENSG00000231856',\n",
       " 'ENSG00000165898',\n",
       " 'ENSG00000110841',\n",
       " 'ENSG00000124177',\n",
       " 'ENSG00000246022',\n",
       " 'ENSG00000165861',\n",
       " 'ENSG00000175106',\n",
       " 'ENSG00000272100',\n",
       " 'ENSG00000187164',\n",
       " 'ENSG00000170890',\n",
       " 'ENSG00000205436',\n",
       " 'ENSG00000135312',\n",
       " 'ENSG00000130635',\n",
       " 'ENSG00000181418',\n",
       " 'ENSG00000227165',\n",
       " 'ENSG00000148426',\n",
       " 'ENSG00000278153',\n",
       " 'ENSG00000272599',\n",
       " 'ENSG00000092978',\n",
       " 'ENSG00000065325',\n",
       " 'ENSG00000038219',\n",
       " 'ENSG00000102921',\n",
       " 'ENSG00000188206',\n",
       " 'ENSG00000179965',\n",
       " 'ENSG00000006451',\n",
       " 'ENSG00000242515',\n",
       " 'ENSG00000132185',\n",
       " 'ENSG00000280132',\n",
       " 'ENSG00000270002',\n",
       " 'ENSG00000249375',\n",
       " 'ENSG00000143224',\n",
       " 'ENSG00000280276',\n",
       " 'ENSG00000254703',\n",
       " 'ENSG00000105778',\n",
       " 'ENSG00000130584',\n",
       " 'ENSG00000073417',\n",
       " 'ENSG00000268568',\n",
       " 'ENSG00000178773',\n",
       " 'ENSG00000267673',\n",
       " 'ENSG00000255663',\n",
       " 'ENSG00000070756',\n",
       " 'ENSG00000112699',\n",
       " 'ENSG00000105193',\n",
       " 'ENSG00000198642',\n",
       " 'ENSG00000281856',\n",
       " 'ENSG00000213279',\n",
       " 'ENSG00000244476',\n",
       " 'ENSG00000119684',\n",
       " 'ENSG00000100014',\n",
       " 'ENSG00000137414',\n",
       " 'ENSG00000163273',\n",
       " 'ENSG00000186971',\n",
       " 'ENSG00000243896',\n",
       " 'ENSG00000129250',\n",
       " 'ENSG00000178372',\n",
       " 'ENSG00000235029',\n",
       " 'ENSG00000255595',\n",
       " 'ENSG00000138668',\n",
       " 'ENSG00000145782',\n",
       " 'ENSG00000102452',\n",
       " 'ENSG00000167523',\n",
       " 'ENSG00000105989',\n",
       " 'ENSG00000169994',\n",
       " 'ENSG00000272950',\n",
       " 'ENSG00000263345',\n",
       " 'ENSG00000166669',\n",
       " 'ENSG00000224985',\n",
       " 'ENSG00000231937',\n",
       " 'ENSG00000110713',\n",
       " 'ENSG00000188163',\n",
       " 'ENSG00000078124',\n",
       " 'ENSG00000103249',\n",
       " 'ENSG00000086589',\n",
       " 'ENSG00000140564',\n",
       " 'ENSG00000142528',\n",
       " 'ENSG00000272416',\n",
       " 'ENSG00000273141',\n",
       " 'ENSG00000279212',\n",
       " 'ENSG00000150773',\n",
       " 'ENSG00000162728',\n",
       " 'ENSG00000260918',\n",
       " 'ENSG00000100429',\n",
       " 'ENSG00000090554',\n",
       " 'ENSG00000142959',\n",
       " 'ENSG00000115568',\n",
       " 'ENSG00000170776',\n",
       " 'ENSG00000028137',\n",
       " 'ENSG00000185436',\n",
       " 'ENSG00000102858',\n",
       " 'ENSG00000121742',\n",
       " 'ENSG00000270589',\n",
       " 'ENSG00000139988',\n",
       " 'ENSG00000243244',\n",
       " 'ENSG00000150687',\n",
       " 'ENSG00000152454',\n",
       " 'ENSG00000185686',\n",
       " 'ENSG00000070388',\n",
       " 'ENSG00000186185',\n",
       " 'ENSG00000197841',\n",
       " 'ENSG00000017260',\n",
       " 'ENSG00000249152',\n",
       " 'ENSG00000241860',\n",
       " 'ENSG00000230724',\n",
       " 'ENSG00000268205',\n",
       " 'ENSG00000069869',\n",
       " 'ENSG00000261773',\n",
       " 'ENSG00000100324',\n",
       " 'ENSG00000089199',\n",
       " 'ENSG00000102024',\n",
       " 'ENSG00000111344',\n",
       " 'ENSG00000275342',\n",
       " 'ENSG00000155275',\n",
       " 'ENSG00000249992',\n",
       " 'ENSG00000279246',\n",
       " 'ENSG00000164405',\n",
       " 'ENSG00000169583',\n",
       " 'ENSG00000103591',\n",
       " 'ENSG00000237862',\n",
       " 'ENSG00000095066',\n",
       " 'ENSG00000170468',\n",
       " 'ENSG00000277373',\n",
       " 'ENSG00000186522',\n",
       " 'ENSG00000101493',\n",
       " 'ENSG00000169126',\n",
       " 'ENSG00000230091',\n",
       " 'ENSG00000277767',\n",
       " 'ENSG00000186847',\n",
       " 'ENSG00000090612',\n",
       " 'ENSG00000142765',\n",
       " 'ENSG00000197063',\n",
       " 'ENSG00000260641',\n",
       " 'ENSG00000140323',\n",
       " 'ENSG00000236751',\n",
       " 'ENSG00000167792',\n",
       " 'ENSG00000278952',\n",
       " 'ENSG00000145293',\n",
       " 'ENSG00000162944',\n",
       " 'ENSG00000130119',\n",
       " 'ENSG00000157181',\n",
       " 'ENSG00000186638',\n",
       " 'ENSG00000228021',\n",
       " 'ENSG00000188001',\n",
       " 'ENSG00000126467',\n",
       " 'ENSG00000267508',\n",
       " 'ENSG00000103351',\n",
       " 'ENSG00000082269',\n",
       " 'ENSG00000143473',\n",
       " 'ENSG00000163214',\n",
       " 'ENSG00000077782',\n",
       " 'ENSG00000234614',\n",
       " 'ENSG00000101158',\n",
       " 'ENSG00000186710',\n",
       " 'ENSG00000273174',\n",
       " 'ENSG00000068878',\n",
       " 'ENSG00000123737',\n",
       " 'ENSG00000127483',\n",
       " 'ENSG00000107959',\n",
       " 'ENSG00000279070',\n",
       " 'ENSG00000124207',\n",
       " 'ENSG00000097046',\n",
       " 'ENSG00000105127',\n",
       " 'ENSG00000101901',\n",
       " 'ENSG00000173681',\n",
       " 'ENSG00000187010',\n",
       " 'ENSG00000106366',\n",
       " 'ENSG00000237883',\n",
       " 'ENSG00000073331',\n",
       " 'ENSG00000009954',\n",
       " 'ENSG00000101452',\n",
       " 'ENSG00000112419',\n",
       " 'ENSG00000236054',\n",
       " 'ENSG00000137103',\n",
       " 'ENSG00000206052',\n",
       " 'ENSG00000167674',\n",
       " 'ENSG00000183324',\n",
       " 'ENSG00000111704',\n",
       " 'ENSG00000161920',\n",
       " 'ENSG00000276533',\n",
       " 'ENSG00000250590',\n",
       " 'ENSG00000165660',\n",
       " 'ENSG00000224086',\n",
       " 'ENSG00000188994',\n",
       " 'ENSG00000250303',\n",
       " 'ENSG00000126432',\n",
       " 'ENSG00000231704',\n",
       " 'ENSG00000233009',\n",
       " 'ENSG00000253125',\n",
       " 'ENSG00000262089',\n",
       " 'ENSG00000197714',\n",
       " 'ENSG00000274654',\n",
       " 'ENSG00000092020',\n",
       " 'ENSG00000196468',\n",
       " 'ENSG00000116977',\n",
       " 'ENSG00000189334',\n",
       " 'ENSG00000173988',\n",
       " 'ENSG00000272908',\n",
       " 'ENSG00000141338',\n",
       " 'ENSG00000143891',\n",
       " 'ENSG00000102383',\n",
       " 'ENSG00000280211',\n",
       " 'ENSG00000124743',\n",
       " 'ENSG00000183671',\n",
       " 'ENSG00000279394',\n",
       " 'ENSG00000236668',\n",
       " 'ENSG00000138495',\n",
       " 'ENSG00000187634',\n",
       " 'ENSG00000269386',\n",
       " 'ENSG00000255655',\n",
       " 'ENSG00000187862',\n",
       " 'ENSG00000161677',\n",
       " 'ENSG00000261159',\n",
       " 'ENSG00000171695',\n",
       " 'ENSG00000271553',\n",
       " 'ENSG00000204580',\n",
       " 'ENSG00000260123',\n",
       " 'ENSG00000181513',\n",
       " 'ENSG00000184436',\n",
       " 'ENSG00000123609',\n",
       " 'ENSG00000130653',\n",
       " 'ENSG00000170689',\n",
       " 'ENSG00000254553',\n",
       " 'ENSG00000172247',\n",
       " 'ENSG00000269564',\n",
       " 'ENSG00000197245',\n",
       " 'ENSG00000100344',\n",
       " 'ENSG00000197989',\n",
       " 'ENSG00000196843',\n",
       " 'ENSG00000164053',\n",
       " 'ENSG00000282221',\n",
       " 'ENSG00000162959',\n",
       " 'ENSG00000120925',\n",
       " 'ENSG00000107863',\n",
       " 'ENSG00000225434',\n",
       " 'ENSG00000140521',\n",
       " 'ENSG00000205143',\n",
       " 'ENSG00000144061',\n",
       " 'ENSG00000131323',\n",
       " 'ENSG00000081803',\n",
       " 'ENSG00000123106',\n",
       " 'ENSG00000165379',\n",
       " 'ENSG00000105552',\n",
       " 'ENSG00000118004',\n",
       " 'ENSG00000214021',\n",
       " 'ENSG00000165816',\n",
       " 'ENSG00000267645',\n",
       " 'ENSG00000161914',\n",
       " 'ENSG00000225173',\n",
       " 'ENSG00000105171',\n",
       " 'ENSG00000229921',\n",
       " 'ENSG00000124535',\n",
       " 'ENSG00000231720',\n",
       " 'ENSG00000240015',\n",
       " 'ENSG00000163923',\n",
       " 'ENSG00000234444',\n",
       " 'ENSG00000167264',\n",
       " 'ENSG00000272482',\n",
       " 'ENSG00000245975',\n",
       " 'ENSG00000196431',\n",
       " 'ENSG00000115307',\n",
       " 'ENSG00000248714',\n",
       " 'ENSG00000117505',\n",
       " 'ENSG00000154553',\n",
       " 'ENSG00000198467',\n",
       " 'ENSG00000273568',\n",
       " 'ENSG00000059122',\n",
       " 'ENSG00000279836',\n",
       " 'ENSG00000215475',\n",
       " 'ENSG00000172428',\n",
       " 'ENSG00000267060',\n",
       " 'ENSG00000156453',\n",
       " 'ENSG00000225973',\n",
       " 'ENSG00000162493',\n",
       " 'ENSG00000278627',\n",
       " 'ENSG00000132016',\n",
       " 'ENSG00000163689',\n",
       " 'ENSG00000123473',\n",
       " 'ENSG00000137364',\n",
       " 'ENSG00000255501',\n",
       " 'ENSG00000243819',\n",
       " 'ENSG00000267030',\n",
       " 'ENSG00000087274',\n",
       " 'ENSG00000138316',\n",
       " 'ENSG00000159445',\n",
       " 'ENSG00000014257',\n",
       " 'ENSG00000064995',\n",
       " 'ENSG00000248925',\n",
       " 'ENSG00000203688',\n",
       " 'ENSG00000259457',\n",
       " 'ENSG00000236514',\n",
       " 'ENSG00000184923',\n",
       " 'ENSG00000164117',\n",
       " 'ENSG00000100604',\n",
       " 'ENSG00000085788',\n",
       " 'ENSG00000270898',\n",
       " 'ENSG00000225745',\n",
       " 'ENSG00000153560',\n",
       " 'ENSG00000166961',\n",
       " 'ENSG00000173083',\n",
       " 'ENSG00000182782',\n",
       " 'ENSG00000260261',\n",
       " 'ENSG00000182240',\n",
       " 'ENSG00000059377',\n",
       " 'ENSG00000241058',\n",
       " 'ENSG00000167131',\n",
       " 'ENSG00000168404',\n",
       " 'ENSG00000256232',\n",
       " 'ENSG00000279138',\n",
       " 'ENSG00000141433',\n",
       " 'ENSG00000253799',\n",
       " 'ENSG00000270562',\n",
       " 'ENSG00000231185',\n",
       " 'ENSG00000138442',\n",
       " 'ENSG00000173947',\n",
       " 'ENSG00000185973',\n",
       " 'ENSG00000236144',\n",
       " 'ENSG00000113712',\n",
       " 'ENSG00000262011',\n",
       " 'ENSG00000179542',\n",
       " 'ENSG00000149516',\n",
       " 'ENSG00000139636',\n",
       " 'ENSG00000164587',\n",
       " 'ENSG00000104957',\n",
       " 'ENSG00000141574',\n",
       " 'ENSG00000198468',\n",
       " 'ENSG00000237604',\n",
       " 'ENSG00000184571',\n",
       " 'ENSG00000261116',\n",
       " 'ENSG00000071282',\n",
       " 'ENSG00000266126',\n",
       " 'ENSG00000229852',\n",
       " 'ENSG00000119383',\n",
       " 'ENSG00000164849',\n",
       " 'ENSG00000213023',\n",
       " 'ENSG00000130711',\n",
       " 'ENSG00000279903',\n",
       " 'ENSG00000127837',\n",
       " 'ENSG00000079215',\n",
       " 'ENSG00000006210',\n",
       " 'ENSG00000165863',\n",
       " 'ENSG00000179059',\n",
       " 'ENSG00000095380',\n",
       " 'ENSG00000137463',\n",
       " 'ENSG00000132153',\n",
       " 'ENSG00000058085',\n",
       " 'ENSG00000113430',\n",
       " 'ENSG00000183196',\n",
       " 'ENSG00000171189',\n",
       " 'ENSG00000135828',\n",
       " 'ENSG00000185168',\n",
       " 'ENSG00000175591',\n",
       " 'ENSG00000188629',\n",
       " 'ENSG00000174844',\n",
       " 'ENSG00000277945',\n",
       " 'ENSG00000166111',\n",
       " 'ENSG00000239268',\n",
       " 'ENSG00000263394',\n",
       " 'ENSG00000197355',\n",
       " 'ENSG00000166736',\n",
       " 'ENSG00000205609',\n",
       " 'ENSG00000100206',\n",
       " 'ENSG00000163739',\n",
       " 'ENSG00000224712',\n",
       " 'ENSG00000171848',\n",
       " 'ENSG00000231999',\n",
       " 'ENSG00000103260',\n",
       " 'ENSG00000125844',\n",
       " 'ENSG00000184492',\n",
       " 'ENSG00000163431',\n",
       " 'ENSG00000255021',\n",
       " 'ENSG00000105711',\n",
       " 'ENSG00000277873',\n",
       " 'ENSG00000125970',\n",
       " 'ENSG00000170364',\n",
       " 'ENSG00000198380',\n",
       " 'ENSG00000270074',\n",
       " 'ENSG00000166682',\n",
       " 'ENSG00000104131',\n",
       " 'ENSG00000117569',\n",
       " 'ENSG00000165630',\n",
       " 'ENSG00000143878',\n",
       " 'ENSG00000164252',\n",
       " 'ENSG00000251138',\n",
       " 'ENSG00000262001',\n",
       " 'ENSG00000101384',\n",
       " 'ENSG00000135740',\n",
       " 'ENSG00000124171',\n",
       " 'ENSG00000137824',\n",
       " 'ENSG00000112531',\n",
       " 'ENSG00000228878',\n",
       " 'ENSG00000182508',\n",
       " 'ENSG00000222047',\n",
       " 'ENSG00000204228',\n",
       " 'ENSG00000163527',\n",
       " 'ENSG00000099904',\n",
       " 'ENSG00000269911',\n",
       " 'ENSG00000107736',\n",
       " 'ENSG00000184271',\n",
       " 'ENSG00000198814',\n",
       " 'ENSG00000060558',\n",
       " 'ENSG00000186377',\n",
       " 'ENSG00000180098',\n",
       " 'ENSG00000119242',\n",
       " 'ENSG00000142494',\n",
       " 'ENSG00000116704',\n",
       " 'ENSG00000170836',\n",
       " 'ENSG00000274460',\n",
       " 'ENSG00000261195',\n",
       " 'ENSG00000256925',\n",
       " 'ENSG00000270189',\n",
       " 'ENSG00000125249',\n",
       " 'ENSG00000124802',\n",
       " 'ENSG00000180592',\n",
       " 'ENSG00000254639',\n",
       " 'ENSG00000136682',\n",
       " 'ENSG00000227619',\n",
       " 'ENSG00000115687',\n",
       " 'ENSG00000183283',\n",
       " ...}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atlas_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "A2dzCKGQzTYX",
   "metadata": {
    "id": "A2dzCKGQzTYX"
   },
   "source": [
    "\n",
    "\n",
    "We keep transcripts with the following combinations of GenCode annotation:\n",
    "\n",
    "    gene_type=transcript_type=lncRNA\n",
    "    gene_type=transcript_type=protein_coding\n",
    "\n",
    "The goal is to avoid these GenCode transcript types:\n",
    "\n",
    "    transcript_type=retained_intron\n",
    "    transcript_type=protein_coding_CDS_not_defined\n",
    "    transcript_type=protein_coding_LoF # note the bug-inducing partial string match\n",
    "    transcript_type=nonsense_mediated_decay\n",
    "    transcript_type=non_stop_decay\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "tt0a3Fe_zQbA",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "tt0a3Fe_zQbA",
    "outputId": "0e36f8a1-9b13-4274-d00a-918288dd7374"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:15:08.936001\n",
      "Gencode good pc transcripts pc/nc: 89305\n",
      "Gencode good nc transcripts pc/nc: 56049\n"
     ]
    }
   ],
   "source": [
    "def load_annotated_transcripts(filepath):\n",
    "    pc_tids = set()\n",
    "    nc_tids = set()\n",
    "    with open (filepath,'r') as handle:\n",
    "        for row in handle:\n",
    "            columns = row.split('\\t')\n",
    "            # Avoid comment lines\n",
    "            if len(columns)>=9 and columns[2] == 'transcript':\n",
    "                comments = columns[8]\n",
    "                pairs = comments.split(';')\n",
    "                tid = None\n",
    "                gtype = None\n",
    "                ttype = None\n",
    "                for pair in pairs:\n",
    "                    if pair.startswith('ID=ENST'):\n",
    "                        tid = pair[3:]\n",
    "                    elif pair.startswith('gene_type='):\n",
    "                        gtype = pair[10:]\n",
    "                    elif pair.startswith('transcript_type='):\n",
    "                        ttype = pair[16:]\n",
    "                if ttype is not None:\n",
    "                    if tid is None:\n",
    "                        raise Exception('transcript type without ID')\n",
    "                    if ttype==gtype:\n",
    "                        if ttype=='protein_coding':\n",
    "                            pc_tids.add(tid)\n",
    "                        elif ttype=='lncRNA':\n",
    "                            nc_tids.add(tid)\n",
    "    return pc_tids, nc_tids\n",
    "print(datetime.now())\n",
    "gencode_pc_transcripts,gencode_nc_transcripts = load_annotated_transcripts(DATA_DIR+ANNOTATION)\n",
    "print('Gencode good pc transcripts pc/nc:', len(gencode_pc_transcripts))\n",
    "print('Gencode good nc transcripts pc/nc:', len(gencode_nc_transcripts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "zbCh3Zz8DX2h",
   "metadata": {
    "id": "zbCh3Zz8DX2h"
   },
   "outputs": [],
   "source": [
    "# gencode_nc_transcripts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "TjdVx1bN1MMG",
   "metadata": {
    "id": "TjdVx1bN1MMG"
   },
   "source": [
    "lncRNA\n",
    "\n",
    "Non-coding genes can halso ave different transcripts called isoforms.\n",
    "\n",
    "Typical defline of the GenCode lncRNA file:\n",
    "\n",
    ">ENST00000456328.2|ENSG00000290825.1|-|OTTHUMT00000362751.1|DDX11L2-202|DDX11L2|1657|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "G4g5S1i2MBs8",
   "metadata": {
    "id": "G4g5S1i2MBs8"
   },
   "outputs": [],
   "source": [
    "# GenCode inputs\n",
    "\n",
    "ANNOTATION = 'gencode.v42.annotation.gff3'\n",
    "NONCODING_SEQUENCE = 'gencode.v42.lncRNA_transcripts.fa'\n",
    "CODING_SEQUENCE = 'gencode.v42.pc_transcripts.fa'\n",
    "# GenCode outputs\n",
    "CODING_CSV = 'gencode.v42.pc_transcripts.csv'\n",
    "NONCODING_CSV = 'gencode.v42.lncRNA_transcripts.csv'\n",
    "# Atlas inputs\n",
    "ATLAS_DIR = \"/content/drive/My Drive/LncRNA/data/LncAtlas_0510/\"\n",
    "ATLAS_FILE = 'lncATLAS_all_data_RCI.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "057jJ9x7QmtO",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "057jJ9x7QmtO",
    "outputId": "9268f6e6-bb81-4da4-d23e-4cf2c179f85d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:15:30.803103\n",
      " Input sequences: 57936\n",
      "Output sequences: 29117\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "infile = DATA_DIR + NONCODING_SEQUENCE\n",
    "outfile = DATA_DIR + NONCODING_CSV\n",
    "converter = fasta_reader(infile,outfile,'lncRNA')\n",
    "converter.allow_these_genes(atlas_genes)\n",
    "converter.allow_these_transcripts(gencode_nc_transcripts)\n",
    "converter.fasta_to_csv()\n",
    "# Without gencode_pc_transcripts filter, Input sequences: 57936 Output sequences: 30139"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "mCGy1tiY1IZG",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "mCGy1tiY1IZG",
    "outputId": "913787f1-daad-4a2e-f99e-f7bea846dca8"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:15:46.732902\n",
      " Input sequences: 111053\n",
      "Output sequences: 85641\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "infile = DATA_DIR + CODING_SEQUENCE\n",
    "outfile = DATA_DIR + CODING_CSV\n",
    "converter = fasta_reader(infile,outfile,'protein_coding')\n",
    "converter.allow_these_genes(atlas_genes)\n",
    "converter.allow_these_transcripts(gencode_pc_transcripts)\n",
    "converter.fasta_to_csv()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2r4zad4Y3RFM",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "2r4zad4Y3RFM",
    "outputId": "f57b7ebc-cb76-4fb3-d375-305581875949"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:17:38.414733\n"
     ]
    }
   ],
   "source": [
    "GENCODE_DIR = \"../../data/GenCode/\"\n",
    "ATLAS_DIR= \"../../data/LncAtlas/\"\n",
    "NONCODING_ALL = 'gencode.v42.lncRNA_transcripts.csv'\n",
    "CODING_ALL = 'gencode.v42.pc_transcripts.csv'\n",
    "ATLAS_DATA='lncATLAS_all_data_RCI.csv'\n",
    "\n",
    "from random import Random\n",
    "from datetime import datetime\n",
    "print(datetime.now())\n",
    "TEST_PORTION = 0.2\n",
    "# Output gene lists\n",
    "CODING_TEST_ID = 'CNRCI_coding_test_genes.gc42.csv'\n",
    "CODING_TRAIN_ID = 'CNRCI_coding_train_genes.gc42.csv'\n",
    "NONCODING_TEST_ID = 'CNRCI_noncoding_test_genes.gc42.csv'\n",
    "NONCODING_TRAIN_ID = 'CNRCI_noncoding_train_genes.gc42.csv'\n",
    "# Output sequence files\n",
    "CODING_TEST_SEQ = 'CNRCI_coding_test_transcripts.gc42.csv'\n",
    "CODING_TRAIN_SEQ = 'CNRCI_coding_train_transcripts.gc42.csv'\n",
    "NONCODING_TEST_SEQ = 'CNRCI_noncoding_test_transcripts.gc42.csv'\n",
    "NONCODING_TRAIN_SEQ = 'CNRCI_noncoding_train_transcripts.gc42.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "D_rSlE-Z3RJL",
   "metadata": {
    "id": "D_rSlE-Z3RJL"
   },
   "outputs": [],
   "source": [
    "def load_sequence_data(filepath):\n",
    "    '''\n",
    "    Load transcript sequences. Also,\n",
    "    Load IDs of the genes for which we have sequence.\n",
    "    The long RNA strings preclude the use of csv.reader utility.\n",
    "    Expect csv file with this header line:\n",
    "    transcript_id,gene_id,biotype,length,sequence\n",
    "    '''\n",
    "    gene_set=set()\n",
    "    sequence_data=[]\n",
    "    with open (filepath) as handle:\n",
    "        header = None\n",
    "        for line in handle:\n",
    "            if header is None:\n",
    "                header = line\n",
    "            else:\n",
    "                line = line.strip()\n",
    "                fields = line.split(',')\n",
    "                transcript_id = fields[0]\n",
    "                gene_id       = fields[1]\n",
    "                biotype       = fields[2]\n",
    "                length        = int(fields[3])\n",
    "                sequence      = fields[4]\n",
    "                if length != len(sequence):\n",
    "                    print(line)\n",
    "                    raise Exception('Lengths do not match')\n",
    "                gene_set.add(gene_id)\n",
    "                sequence_data.append(fields)\n",
    "    gene_list = sorted(list(gene_set))\n",
    "    return gene_list,sequence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "DZSO_SCe3RMr",
   "metadata": {
    "id": "DZSO_SCe3RMr"
   },
   "outputs": [],
   "source": [
    "def inplace_shuffle(rows):\n",
    "    generator = Random()\n",
    "    generator.seed(42)\n",
    "    generator.shuffle(rows)  # in-place\n",
    "\n",
    "def train_test_split(rows):\n",
    "    length = len(rows)\n",
    "    divider = int(length*TEST_PORTION)\n",
    "    train_set = rows[divider:]\n",
    "    test_set = rows[:divider]\n",
    "    return (train_set,test_set)\n",
    "def save_csv(rows,filepath):\n",
    "    with open(filepath,'w') as handle:\n",
    "        header = 'gene_id'\n",
    "        handle.write(header)\n",
    "        handle.write('\\n')\n",
    "        for line in rows:\n",
    "            handle.write(line)\n",
    "            handle.write('\\n')\n",
    "def assert_exclusivity(list1,list2):\n",
    "    set1=set(list1)\n",
    "    set2=set(list2)\n",
    "    if len(set1)!=len(list1) or len(set2)!=len(list2):\n",
    "        raise Exception('Lists contained duplicates')\n",
    "    intersection = set1 & set2\n",
    "    if len(intersection)!=0:\n",
    "        raise Exception('Lists are not exclusive')\n",
    "def get_atlas_genes(filepath):\n",
    "    gene_set = set()\n",
    "    with open (filepath, 'r') as handle:\n",
    "        header = None\n",
    "        for row in handle:\n",
    "            if header is None:\n",
    "                header = row\n",
    "            else:\n",
    "                row = row.strip()\n",
    "                fields = row.split(',')\n",
    "                gene_id = fields[0]\n",
    "                cell_line = fields[1]\n",
    "                data_type = fields[2]\n",
    "                value = fields[3]\n",
    "                if data_type == 'CNRCI' and value != 'NA':\n",
    "                    gene_set.add(gene_id)\n",
    "    return gene_set\n",
    "def get_intersection(all_data, subset_ids):\n",
    "    intersection = []\n",
    "    for gene_id in all_data:\n",
    "        if gene_id in subset_ids:\n",
    "            intersection.append(gene_id)\n",
    "    return intersection\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3mkiqjcc3RPb",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3mkiqjcc3RPb",
    "outputId": "01803b48-9700-4092-d9c0-7566b1229bdc"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:17:49.670508\n",
      "First few coding genes:\n",
      "['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460']\n",
      "First few noncoding genes:\n",
      "['ENSG00000082929', 'ENSG00000099869', 'ENSG00000105501', 'ENSG00000115934', 'ENSG00000116652']\n",
      "2025-05-19 23:17:50.997833\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "coding_genes,    coding_sequence    = load_sequence_data(GENCODE_DIR+CODING_ALL)\n",
    "noncoding_genes, noncoding_sequence = load_sequence_data(GENCODE_DIR+NONCODING_ALL)\n",
    "\n",
    "print('First few coding genes:')\n",
    "print(coding_genes[:5])\n",
    "print('First few noncoding genes:')\n",
    "print(noncoding_genes[:5])\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "V9qDlTb1AQve",
   "metadata": {
    "id": "V9qDlTb1AQve"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "dff = pd.read_csv(GENCODE_DIR+NONCODING_ALL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "TKWhigPhAykT",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 424
    },
    "id": "TKWhigPhAykT",
    "outputId": "132e290b-4865-4b69-d4f7-4806a52a1d90"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>biotype</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>712</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>535</td>\n",
       "      <td>TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2748</td>\n",
       "      <td>CTGATCCATATGAATTCCTCTTATTAAGAAAAATAAAGCATCCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>491</td>\n",
       "      <td>GACAAGTTCGAGCATCTTAAAATGATTCAACAGGAGGAGATAAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>629</td>\n",
       "      <td>GAAGCTCGAGGAAGAGAAAAAAAAACTGGAAGGAGAAATCATAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29112</th>\n",
       "      <td>ENST00000667496.1</td>\n",
       "      <td>ENSG00000229236</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>4474</td>\n",
       "      <td>GCTCTTGTTGCCCATGCTGGAGTGCAGGGGCGCGATCTGCCCGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29113</th>\n",
       "      <td>ENST00000659275.1</td>\n",
       "      <td>ENSG00000229236</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3567</td>\n",
       "      <td>CACCCTGGCCATAGCCCATGTCAGCAGTAACCTATGTCTTTGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29114</th>\n",
       "      <td>ENST00000666666.1</td>\n",
       "      <td>ENSG00000229236</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1469</td>\n",
       "      <td>CTTTGCTAAAAAACGGTCTTCCAGTTTCAGAAGTTCGTGGGTCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29115</th>\n",
       "      <td>ENST00000382764.1</td>\n",
       "      <td>ENSG00000183146</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>878</td>\n",
       "      <td>ACAGGAGGACAAGGACTCAGGGGTCTGCTGGTCCATCTCTGCACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29116</th>\n",
       "      <td>ENST00000399966.9_PAR_Y</td>\n",
       "      <td>ENSG00000185203</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1054</td>\n",
       "      <td>CGGCAGTCACTTTGCGGCCACTATGAGGACTGAGCTTCCTCTCCCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29117 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_id          gene_id biotype  length  \\\n",
       "0            ENST00000473358.1  ENSG00000243485  lncRNA     712   \n",
       "1            ENST00000469289.1  ENSG00000243485  lncRNA     535   \n",
       "2            ENST00000466430.5  ENSG00000238009  lncRNA    2748   \n",
       "3            ENST00000477740.5  ENSG00000238009  lncRNA     491   \n",
       "4            ENST00000471248.1  ENSG00000238009  lncRNA     629   \n",
       "...                        ...              ...     ...     ...   \n",
       "29112        ENST00000667496.1  ENSG00000229236  lncRNA    4474   \n",
       "29113        ENST00000659275.1  ENSG00000229236  lncRNA    3567   \n",
       "29114        ENST00000666666.1  ENSG00000229236  lncRNA    1469   \n",
       "29115        ENST00000382764.1  ENSG00000183146  lncRNA     878   \n",
       "29116  ENST00000399966.9_PAR_Y  ENSG00000185203  lncRNA    1054   \n",
       "\n",
       "                                                sequence  \n",
       "0      GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "1      TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...  \n",
       "2      CTGATCCATATGAATTCCTCTTATTAAGAAAAATAAAGCATCCAGG...  \n",
       "3      GACAAGTTCGAGCATCTTAAAATGATTCAACAGGAGGAGATAAGGA...  \n",
       "4      GAAGCTCGAGGAAGAGAAAAAAAAACTGGAAGGAGAAATCATAGAT...  \n",
       "...                                                  ...  \n",
       "29112  GCTCTTGTTGCCCATGCTGGAGTGCAGGGGCGCGATCTGCCCGTCT...  \n",
       "29113  CACCCTGGCCATAGCCCATGTCAGCAGTAACCTATGTCTTTGTTTT...  \n",
       "29114  CTTTGCTAAAAAACGGTCTTCCAGTTTCAGAAGTTCGTGGGTCATT...  \n",
       "29115  ACAGGAGGACAAGGACTCAGGGGTCTGCTGGTCCATCTCTGCACCT...  \n",
       "29116  CGGCAGTCACTTTGCGGCCACTATGAGGACTGAGCTTCCTCTCCCC...  \n",
       "\n",
       "[29117 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "aQYzt6Cy4ZCZ",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "aQYzt6Cy4ZCZ",
    "outputId": "16fcab9d-cec5-4614-8d6c-4699d0a5ec98"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6270"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(noncoding_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "wLhcpmGd3RSU",
   "metadata": {
    "id": "wLhcpmGd3RSU"
   },
   "outputs": [],
   "source": [
    "atlas_genes = get_atlas_genes(ATLAS_DIR+ATLAS_DATA)\n",
    "coding_genes = get_intersection(coding_genes, atlas_genes)\n",
    "noncoding_genes = get_intersection(noncoding_genes, atlas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hzFIfhjP4doj",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "hzFIfhjP4doj",
    "outputId": "b1e62825-5810-4ee9-9264-18f13a33cba2"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17472"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coding_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "HUJwEsRM3RU_",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HUJwEsRM3RU_",
    "outputId": "7c01fef2-b3ec-41ed-b192-ff26bf660ebf"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:18:53.885857\n",
      "First few coding genes:\n",
      "['ENSG00000107679', 'ENSG00000164647', 'ENSG00000169299', 'ENSG00000156787', 'ENSG00000101464']\n",
      "First few noncoding genes:\n",
      "['ENSG00000185186', 'ENSG00000259005', 'ENSG00000250775', 'ENSG00000266801', 'ENSG00000236352']\n",
      "2025-05-19 23:18:53.892853\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "inplace_shuffle(coding_genes)\n",
    "inplace_shuffle(noncoding_genes)\n",
    "\n",
    "print('First few coding genes:')\n",
    "print(coding_genes[:5])\n",
    "print('First few noncoding genes:')\n",
    "print(noncoding_genes[:5])\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6QvL059-3RYC",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "6QvL059-3RYC",
    "outputId": "0347a1ea-0b36-4aa6-cc66-e6b046487b6c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:18:56.313563\n",
      "First few coding train genes:\n",
      "['ENSG00000212659', 'ENSG00000103269', 'ENSG00000156206', 'ENSG00000106608', 'ENSG00000150627']\n",
      "First few coding test genes:\n",
      "['ENSG00000107679', 'ENSG00000164647', 'ENSG00000169299', 'ENSG00000156787', 'ENSG00000101464']\n",
      "First few noncoding train genes:\n",
      "['ENSG00000213904', 'ENSG00000261766', 'ENSG00000273145', 'ENSG00000259977', 'ENSG00000277463']\n",
      "First few noncoding test genes:\n",
      "['ENSG00000185186', 'ENSG00000259005', 'ENSG00000250775', 'ENSG00000266801', 'ENSG00000236352']\n",
      "2025-05-19 23:18:56.314562\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now())\n",
    "coding_train_set,   coding_test_set    = train_test_split(coding_genes)\n",
    "noncoding_train_set,noncoding_test_set = train_test_split(noncoding_genes)\n",
    "\n",
    "print('First few coding train genes:')\n",
    "print(coding_train_set[:5])\n",
    "print('First few coding test genes:')\n",
    "print(coding_test_set[:5])\n",
    "print('First few noncoding train genes:')\n",
    "print(noncoding_train_set[:5])\n",
    "print('First few noncoding test genes:')\n",
    "print(noncoding_test_set[:5])\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "I1VeXWYS7DWg",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "I1VeXWYS7DWg",
    "outputId": "b708ad21-ba48-41f2-f848-9c4c6e0b645a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-05-19 23:19:01.182267\n",
      "First few coding train genes:\n",
      "['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460']\n",
      "First few coding test genes:\n",
      "['ENSG00000000938', 'ENSG00000000971', 'ENSG00000001036', 'ENSG00000001460', 'ENSG00000001626']\n",
      "First few noncoding train genes:\n",
      "['ENSG00000099869', 'ENSG00000105501', 'ENSG00000116652', 'ENSG00000117242', 'ENSG00000120664']\n",
      "First few noncoding test genes:\n",
      "['ENSG00000082929', 'ENSG00000124915', 'ENSG00000130600', 'ENSG00000145063', 'ENSG00000146666']\n",
      "2025-05-19 23:19:01.186776\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "print(datetime.now())\n",
    "coding_train_sort    = sorted(coding_train_set)\n",
    "coding_test_sort     = sorted(coding_test_set)\n",
    "noncoding_train_sort = sorted(noncoding_train_set)\n",
    "noncoding_test_sort  = sorted(noncoding_test_set)\n",
    "coding_train_set     = None\n",
    "coding_test_set      = None\n",
    "noncoding_train_set  = None\n",
    "noncoding_test_set   = None\n",
    "\n",
    "print('First few coding train genes:')\n",
    "print(coding_train_sort[:5])\n",
    "print('First few coding test genes:')\n",
    "print(coding_test_sort[:5])\n",
    "print('First few noncoding train genes:')\n",
    "print(noncoding_train_sort[:5])\n",
    "print('First few noncoding test genes:')\n",
    "print(noncoding_test_sort[:5])\n",
    "print(datetime.now())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "97gGaE6F7DZL",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "97gGaE6F7DZL",
    "outputId": "f17a1f43-8eb1-4850-e511-2946b94405ed"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../data/LncAtlas/CNRCI_noncoding_test_genes.gc42.csv\n",
      "../../data/LncAtlas/CNRCI_noncoding_train_genes.gc42.csv\n"
     ]
    }
   ],
   "source": [
    "filename = ATLAS_DIR + NONCODING_TEST_ID\n",
    "print(filename)\n",
    "save_csv( noncoding_test_sort, filename)\n",
    "\n",
    "filename = ATLAS_DIR + NONCODING_TRAIN_ID\n",
    "print(filename)\n",
    "save_csv( noncoding_train_sort, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "j49vTSU77Dbi",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "j49vTSU77Dbi",
    "outputId": "2f7fc6ad-d558-4a36-d122-7ec72330944b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Noncoding total, filtered, train, test: 5827 4662 1165\n"
     ]
    }
   ],
   "source": [
    "print('Noncoding total, filtered, train, test:',\n",
    "      len(noncoding_genes), len(noncoding_train_sort), len(noncoding_test_sort))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3kh3zYBf7DeA",
   "metadata": {
    "id": "3kh3zYBf7DeA"
   },
   "outputs": [],
   "source": [
    "def save_seq(gene_list,sequence_data,filepath):\n",
    "    with open(filepath,'w') as handle:\n",
    "        header = 'transcript_id,gene_id,biotype,length,sequence'\n",
    "        handle.write(header)\n",
    "        handle.write('\\n')\n",
    "        valid_ids = set(gene_list)\n",
    "        for fields in sequence_data:\n",
    "            transcript_id = fields[0]\n",
    "            gene_id       = fields[1]\n",
    "            biotype       = fields[2]\n",
    "            length        = fields[3]\n",
    "            sequence      = fields[4]\n",
    "            if gene_id in valid_ids:\n",
    "                line = ','.join(fields)\n",
    "                handle.write(line)\n",
    "                handle.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "WUKrm6Q97Djm",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "WUKrm6Q97Djm",
    "outputId": "395c3da0-a028-42ea-e116-c03d78093fff"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Save transcript sequences\n",
      "../../data/LncAtlas/CNRCI_coding_test_transcripts.gc42.csv\n",
      "../../data/LncAtlas/CNRCI_coding_train_transcripts.gc42.csv\n",
      "../../data/LncAtlas/CNRCI_noncoding_test_transcripts.gc42.csv\n",
      "../../data/LncAtlas/CNRCI_noncoding_train_transcripts.gc42.csv\n"
     ]
    }
   ],
   "source": [
    "print('Save transcript sequences')\n",
    "filename = ATLAS_DIR + CODING_TEST_SEQ\n",
    "print(filename)\n",
    "save_seq( coding_test_sort, coding_sequence, filename)\n",
    "\n",
    "filename = ATLAS_DIR + CODING_TRAIN_SEQ\n",
    "print(filename)\n",
    "save_seq( coding_train_sort, coding_sequence, filename)\n",
    "\n",
    "filename = ATLAS_DIR + NONCODING_TEST_SEQ\n",
    "print(filename)\n",
    "save_seq( noncoding_test_sort, noncoding_sequence, filename)\n",
    "\n",
    "filename = ATLAS_DIR + NONCODING_TRAIN_SEQ\n",
    "print(filename)\n",
    "save_seq( noncoding_train_sort, noncoding_sequence, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "Nn3WGa9m7Dwe",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Nn3WGa9m7Dwe",
    "outputId": "6d6ac800-1d60-4e96-a251-83453b19e91f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n",
      "2025-05-19 23:19:43.375474\n"
     ]
    }
   ],
   "source": [
    "print('done')\n",
    "print(datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16e31f70",
   "metadata": {
    "id": "16e31f70"
   },
   "source": [
    "## mRNA\n",
    "\n",
    "The cds file contains the coding sequence, that is, the processed transcript, which is the gene sequence minus the untranslated UTR and introns.\n",
    "\n",
    "Genes can have different transcripts called isoforms.\n",
    "\n",
    "Some protein_coding genes have degenerate, non-coding transcripts. For example, a protein_coding gene might have one transcript marked for nonsense-mediated_decay because it is ill-formed.\n",
    "\n",
    "Here, we filter for the protein_coding transcripts only.\n",
    "\n",
    "Typical defline of cds file:\n",
    "ENST00000631435.1 cds chromosome:GRCh38:CHR_HSCHR7_2_CTG6:142847306:142847317:1 gene:ENSG00000282253.1 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:TRBD1 description:T cell receptor beta diversity 1 [Source:HGNC Symbol;Acc:HGNC:12158]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f19d8702",
   "metadata": {
    "id": "f19d8702"
   },
   "source": [
    "## lncRNA\n",
    "\n",
    "The ncrna file contains non-coding sequence, that is, transcripts from non-coding genes.\n",
    "\n",
    "Non-coding genes can have different transcripts called isoforms.\n",
    "\n",
    "Some non-coding genes have degenerate transcripts. For example, a lncRNA gene might have one transcript marked retained_intron because it was not processed correctly.\n",
    "\n",
    "Here, we filter for the lncRNA transcripts only.\n",
    "\n",
    "Typical defline of ncrna file:\n",
    ">ENST00000516993.1 ncrna chromosome:GRCh38:1:26593940:26594041:-1 gene:ENSG00000252802.1 gene_biotype:misc_RNA transcript_biotype:misc_RNA gene_symbol:Y_RNA description:Y RNA [Source:RFAM;Acc:RF00019]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "uksAO6nb6rnr",
   "metadata": {
    "id": "uksAO6nb6rnr"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
